keps/sig-auth/0000-authority-delegation.md

```diff
@@ -0,0 +1,259 @@
+acts should be explicit. This prevents a controller from misusing permission
+granted ambiently to the controller.
+
+Authority is selected per request by passing an “Authority-Selector” header
+with the request:
+
+```
+Authority-Selector: {“authorityReference”:{“kind”:”ClusterRole”,”name”:”viewer”}}
+```
+or:
+```
+Authority-Selector: {“authorityReference”:{“kind”:”AuthorityDelegation”,”uid”:”0000-0...”}}
+```
+
+The value of the selector is a `SelectedAuthority` object in the
+`authorization.k8s.io` API group:
+
+```go
+type SelectedAuthority struct {
+  metav1.TypeMeta
+  AuthorityReference AuthorityReference
+}
+
+type AuthorityReference struct {
+  metav1.TypeMeta
+  Name string
+  UID types.UID
+}
+```
+
+When a request is received by an API, the API would normally iterate over the
+set of authorizers and each authorizer would check all policies that would
+permit the action, searching for a policy that permits the action and is bound
+to the user.
+
+When a request is received by the API with a selected authority, the API would
+check the specific authorizer that supports the AuthorityReference kind, and
+the authorizer would check the request against the specific policy referenced
+in by the SelectedAuthority.
+
+### Persistent Authority Delegation
+
+#### AuthorityDelegation
+
+A new non-namespaced persistent object will be added to the
+`authorization.k8s.io` API group to store instances of authority delegation.
+
+```go
+type AuthoritiyDelegation struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  // The user that created the delegation.
+  User Info
+  // The principal that is delegating authority.
+  Subject SubjectRef
+  // The principals that are delegated authority.
+  Delegates []SubjectRef
+  // defined above
+  Authority AuthorityReference
+  // If this delegation is created selecting a previously delegated authority,
+  // the delegation chain is recorded here.
+  DelegationHistory []AuthorityDelegationReference
+}
+
+type SubjectRef struct {
+  Kind SubjectKind
+  Name string
+  UID types.UID
+}
+
+const (
+  GroupKind SubjectKind = "Group"
+  UserKind SubjectKind = "User"
+)
+
+type AuthrityDelegationReference struct {
+  UID types.UID
+}
+```
+
+#### AuthorityDelegationPolicy
+
+A new policy will be introduced to implictly create `AuthorityDelegations` in
+reaction to API actions. The `AuthorityDelegationPolicy` will be a
+non-namespaced object in the `authorization.k8s.io` API group.
+
+```go
+type AuthorityDelegationPolicy struct {
+  // The resource attributes of an action to implicitly create a delegation for.
+  ResourceAttributes *ResourceAttributes
+  // The principals that are delegated authority.
+  Delegates []SubjectRef
+  // The permissions that are delegated.
+  Authority AuthorityReference
+}
+```
+
+When an API action matching an `AuthorityDelegationPolicy` is processed, an
+`AuthorityDelegation` is implicitly created before any objects are persisted.
+
+#### Example of a Cascading Delegation
```

Comment:
What about non-create actions (e.g. update, delete)? It seems like this proposal only addresses the create case.